diff --git a/../mpv/player/lua/osc.lua b/osc.lua
index 6132aa9..5c73229 100644
--- a/osc.lua
+++ b/osc.lua
@@ -1711,6 +1711,13 @@ local function bar_layout(direction, slim)
     lo.geometry = geo
     lo.style = osc_styles.smallButtonsBar
 
+    -- START quality-menu
+    geo = { x = geo.x - geo.w - padX, y = geo.y, an = geo.an, w = geo.w, h = geo.h }
+    lo = add_layout("quality-menu")
+    lo.geometry = geo
+    lo.style = osc_styles.smallButtonsBar
+    -- END quality-menu
+
     -- Volume
     geo = { x = geo.x - geo.w - padX, y = geo.y, an = geo.an, w = geo.w, h = geo.h }
     lo = add_layout("volume")
@@ -2136,6 +2143,17 @@ local function osc_init()
             string.format("%3.0fs", sec))
     end
 
+    -- START quality-menu
+    ne = new_element("quality-menu", "button")
+    ne.content = function()
+        return ("≚")
+    end
+    ne.eventresponder["mbtn_left_up"] =
+        function () mp.commandv("script-message", "video_formats_toggle") end
+    ne.eventresponder["mbtn_right_up"] =
+        function () mp.commandv("script-message", "audio_formats_toggle") end
+    -- END quality-menu
+
     -- volume
     ne = new_element("volume", "button")
 
